// ng-table-virtual-scroll/esm2020/lib/table-virtual-scroll.module.mjs
import { NgModule } from '@angular/core';
import { TableItemSizeDirective } from './table-item-size.directive';
import * as i0 from "@angular/core";
export class TableVirtualScrollModule {
}
TableVirtualScrollModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: TableVirtualScrollModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TableVirtualScrollModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: TableVirtualScrollModule, declarations: [TableItemSizeDirective], exports: [TableItemSizeDirective] });
TableVirtualScrollModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: TableVirtualScrollModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.8", ngImport: i0, type: TableVirtualScrollModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TableItemSizeDirective],
                    imports: [],
                    exports: [TableItemSizeDirective]
                }]
        }] });
